Add support for Microsoft Azure Storage Blob in ObjectStore

lizardoluis opened the pull request:
Adds support for Microsoft Azure Blob Storage to the object store abstraction in the project. It introduces Azure as storage backend, updates configuration parsing and builder logic, and provides thorough tests for Azure bucket detection and configuration. The changes ensure Azure can be used alongside AWS S3 and Google Cloud Storage with similar APIs and configuration patterns.

lizardoluis committed:
Add support for Microsoft Azure in ObjectStore

JanKaul commented on iceberg-rust/src/object_store/mod.rs:
```diff
@@ -125,6 +151,9 @@ impl ObjectStoreBuilder {
+            }
             (ObjectStoreBuilder::S3(aws), ConfigKey::AWS(key)) => {
                 ObjectStoreBuilder::S3(Box::new(aws.with_config(key, value)))
             }
```
I realized that it's not good to parse the config string into a `ConfigKey` before passing it to the builder, because there is no knowledge about which provider(s3,gcs,azure) to parse it for. This is an issue if the providers have matching config keys. That's why one of the tests is failing.
Instead we need to pass the string to `with_config()` and then parse the `ConfigKey` inside of it when we have the appropriate information.
I'll create a PR to fix this.

JanKaul replied:
https://github.com/splitgraph/iceberg-rust/pull/20

lizardoluis replied:
In our system, we use the `ObjectStoreScheme::parse()` with the URL to define which builder to use.

JanKaul replied:
Do you use the `with_config` method? Or do you build the entire Builder beforehand and then pass it to `ObjectStoreBuilder`?

lizardoluis replied:
Both ways. We have some fixed options that we pass to the builder in a structured way, but the user can set more and those we add with `with_config()`.

JanKaul committed:
fix object-store with_config

JanKaul commented on iceberg-rust/src/object_store/mod.rs:
```diff
@@ -93,6 +112,9 @@ pub enum ConfigKey {
 impl FromStr for ConfigKey {
     type Err = object_store::Error;
     fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Ok(x) = s.parse() {
+            return Ok(ConfigKey::Azure(x));
+        };
         if let Ok(x) = s.parse() {
             return Ok(ConfigKey::AWS(x));
         };
```
Could you swap the order of Azure and AWS? Because then the tests shouldn't fail anymore.

This however could lead to problems on your side. That's why I'm proposing to update the `with_config` method. Which should fix the issues in either case.

The problem here is that the output ConfigKey depends on the order of the which the providers are parsed. A keywork like "endpoint" which is supported by multiple providers, will always be parsed as a key of the provider that is listed here first. Even though it might want to be used for a different provider.

lizardoluis replied:
I can do. Should I pick your changes on the other PR? Or are you going to merge them? 
I also found a problem with the way I'm extracting the buckets. I'll fix that first and set this PR as draft until then

lizardoluis replied:
Ah, just noticed that your PR is on our fork. I'll merge it

JanKaul replied:
I hope it doesn't cause any problems on your side but I would like to inlcude my PR. This should make it more robust when using config string with different providers.

When my PR is merged, you also don't have to swap the order, as this is handled automatically.

lizardoluis committed:
Fix object-store with_config

This commit moves the parsing of the config keys into the with_config method of the ObjectStoreBuilder. This is to avoid issues where multiple providers use the same config keys and it hence can lead to a mismatch between ObjectStoreBuilder and ConfigKey

JanKaul commented:
Looks good to me now

lizardoluis commented:
Ok, I didn’t add any other changes — it can be merged as is. Thanks for handling your side of the fix.

I marked it as a draft because I wasn’t sure about the [bucket parser](https://github.com/JanKaul/iceberg-rust/pull/271/files#diff-3960679c2064ce8c0ccc46a469f25ccf502a37dfa099cfb49bb6cd21347e392aR70-R80).

Azure uses a different URL format (`https://myaccount.blob.core.windows.net/mycontainer/myblob`), so I wasn’t sure whether the parser should return just `mycontainer` (as I implemented) or the full path `myaccount.blob.core.windows.net/mycontainer`.

JanKaul commented:
I don't have a lot of experience with Azure. Well, we can just fix it in another PR if the need arises.

Thanks for the PR!
